vllm/model_executor/models/phi_1_5.py

```diff
@@ -0,0 +1,286 @@
+        super().__init__()
+        self.ln = nn.LayerNorm(config.n_embd, eps=config.layer_norm_epsilon)
+        self.mixer = PhiAttention(config)
+        self.mlp = PhiMLP(config)
+
+    def forward(
+        self,
+        position_ids: torch.Tensor,
+        hidden_states: torch.Tensor,
+        kv_cache: KVCache,
+        input_metadata: InputMetadata,
+        cache_event: Optional[torch.cuda.Event],
+    ) -> torch.Tensor:
+        residual = hidden_states
+        hidden_states = self.ln(hidden_states)
+        attn_outputs = self.mixer(
+            position_ids=position_ids,
+            hidden_states=hidden_states,
+            kv_cache=kv_cache,
+            input_metadata=input_metadata,
+            cache_event=cache_event,
+        )
+        feed_forward_hidden_states = self.mlp(hidden_states)
+        hidden_states = attn_outputs + feed_forward_hidden_states + residual
+        return hidden_states
+
+
+class PhiCausalLMHead(nn.Module):
+
+    def __init__(self, config: PretrainedConfig):
+        super().__init__()
+        self.ln = nn.LayerNorm(config.n_embd, eps=config.layer_norm_epsilon)
+        self.linear = ColumnParallelLinear(
+            config.n_embd,
+            config.vocab_size,
+            gather_output=False,
+        )
+        self.sampler = Sampler(config.vocab_size)
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        input_metadata: InputMetadata,
+    ):
+        hidden_states = self.ln(hidden_states)
+        next_tokens = self.sampler(self.linear.weight, hidden_states,
+                                   input_metadata, self.linear.bias)
+        return next_tokens
+
+
+class PhiForCausalLM(nn.Module):
+
+    def __init__(self, config: PretrainedConfig):
+        super().__init__()
+        self.config = config
+        modules = [PhiEmbedding(config)]
+        modules += [PhiLayer(config) for _ in range(config.n_layer)]
+        modules.append(PhiCausalLMHead(config))
+        self.layers = nn.Sequential(*modules)
+
+    def forward(
+        self,
+        input_ids: torch.Tensor,
+        positions: torch.Tensor,
+        kv_caches: List[KVCache],
+        input_metadata: InputMetadata,
+        cache_events: Optional[List[torch.cuda.Event]],
+    ) -> SamplerOutput:
+        hidden_states = self.layers[0](input_ids)
+        for i in range(self.config.n_layer):
+            if cache_events is None:
+                cache_event = None
+            else:
+                cache_event = cache_events[i]
+            layer = self.layers[i + 1]
+            hidden_states = layer(
+                positions,
+                hidden_states,
+                kv_caches[i],
+                input_metadata,
+                cache_event,
+            )
+        lm_logits = self.layers[-1](hidden_states, input_metadata)
+        return lm_logits
+
+    _column_parallel_weights = [
+        "wte.weight", "linear.weight", "linear.bias", "fc1.weight", "fc1.bias"
+    ]
```

Comment:
It seems TP is not supported well: 1) "Wqkv.weight" and "Wqkv.bias" are missing here, and 2) even after adding the two, the model generates gibberish outputs. Could you double check?

Reply:
1. Yeah, will fix it 
2. Hm, interesting. Last time I tried running tests, it was okey with fp32. Could you please provide the prompt that you used?

Reply:
Btw, for 1. I took my inspiration from the GPT-NeoX implementation, perhaps there is the same bug in [the implementation](https://github.com/vllm-project/vllm/blob/660a7fcfa40d62305ecba6bc6352c4026d56d680/vllm/model_executor/models/gpt_neox.py#L239)

Reply:
I've run `examples/llm_engine_example.py`. The output looks good when TP is 1, but not when TP is > 1.

* TP = 1
```
$ # python examples/llm_engine_example.py --model microsoft/phi-1_5 --trust-remote-code -tp 1
INFO 11-15 22:23:35 llm_engine.py:72] Initializing an LLM engine with config: model='microsoft/phi-1_5', tokenizer='microsoft/phi-1_5', tokenizer_mode=auto, revision=None, tokenizer_revision=None, trust_remote_code=True, dtype=torch.float16, max_seq_len=2048, download_dir=None, load_format=auto, tensor_parallel_size=1, quantization=None, seed=0)
INFO 11-15 22:23:46 llm_engine.py:207] # GPU blocks: 5602, # CPU blocks: 1365
INFO 11-15 22:23:49 llm_engine.py:624] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 0.0 tokens/s, Running: 1 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 0.0%, CPU KV cache usage: 0.0%
RequestOutput(request_id=0, prompt='A robot may not injure a human being', prompt_token_ids=[32, 9379, 743, 407, 42206, 257, 1692, 852], prompt_logprobs=[None, {9379: -9.672202110290527, 25: -0.3518897294998169}, {743: -4.376789569854736, 318: -1.3142896890640259}, {407: -3.103748321533203, 307: -1.3068734407424927}, {42206: -10.088486671447754, 307: -0.8150489330291748}, {257: -1.313309669494629}, {1692: -0.9726755619049072}, {852: -4.1470537185668945, 611: -1.4126787185668945}], outputs=[CompletionOutput(index=0, text=' if it is programmed to avoid certain areas or objects.\n\nExercise 3', token_ids=[611, 340, 318, 27402, 284, 3368, 1728, 3006, 393, 5563, 13, 198, 198, 3109, 23697, 513], cumulative_logprob=-14.929404081776738, logprobs=[{611: -1.6466339826583862}, {340: -0.2971111536026001}, {318: -1.2996068000793457}, {27402: -0.26270025968551636}, {284: -0.11931092292070389}, {3368: -0.6281225681304932}, {1728: -2.2663183212280273}, {3006: -1.0307843685150146}, {393: -0.4654202163219452}, {5563: -1.5459181070327759}, {13: -0.887886643409729}, {198: -0.61713045835495}, {198: -0.9203287363052368}, {3109: -1.65543794631958}, {23697: -0.018604421988129616}, {513: -1.2680891752243042}], finish_reason=length)], finished=True)
RequestOutput(request_id=1, prompt='To be or not to be,', prompt_token_ids=[2514, 307, 393, 407, 284, 307, 11], prompt_logprobs=None, outputs=[CompletionOutput(index=0, text=' that is the question.\n\nIn the midst of all this chaos, there', token_ids=[326, 318, 262, 1808, 13, 198, 198, 818, 262, 15925, 286, 477, 428, 11918, 11, 612], cumulative_logprob=-12.005558035519243, logprobs=None, finish_reason=length)], finished=True)
RequestOutput(request_id=2, prompt='What is the meaning of life?', prompt_token_ids=[2061, 318, 262, 3616, 286, 1204, 30], prompt_logprobs=None, outputs=[CompletionOutput(index=0, text='\nAnswer: This is a philosophical question that has been debated for centuries. Some', token_ids=[198, 33706, 25, 770, 318, 257, 17580, 1808, 326, 468, 587, 24594, 329, 10675, 13, 2773], cumulative_logprob=-5.963053565472364, logprobs=None, finish_reason=length), CompletionOutput(index=4, text='\nAnswer: This is a philosophical question that has puzzled humans for centuries.\n', token_ids=[198, 33706, 25, 770, 318, 257, 17580, 1808, 326, 468, 36907, 5384, 329, 10675, 13, 198], cumulative_logprob=-6.1559734754264355, logprobs=None, finish_reason=length)], finished=True)
RequestOutput(request_id=3, prompt='It is only with the heart that one can see rightly', prompt_token_ids=[1026, 318, 691, 351, 262, 2612, 326, 530, 460, 766, 22956], prompt_logprobs=None, outputs=[CompletionOutput(index=1, text='; with the ear that one can hear rightly; with the tongue that one can', token_ids=[26, 351, 262, 1027, 326, 530, 460, 3285, 22956, 26, 351, 262, 11880, 326, 530, 460], cumulative_logprob=-5.308803586587601, logprobs=None, finish_reason=length), CompletionOutput(index=2, text='; with the ear that one can hear rightly; with the mind that one can', token_ids=[26, 351, 262, 1027, 326, 530, 460, 3285, 22956, 26, 351, 262, 2000, 326, 530, 460], cumulative_logprob=-6.430019140192599, logprobs=None, finish_reason=length), CompletionOutput(index=0, text='; with the ear that one can hear rightly; and with the tongue that one', token_ids=[26, 351, 262, 1027, 326, 530, 460, 3285, 22956, 26, 290, 351, 262, 11880, 326, 530], cumulative_logprob=-7.173643220390659, logprobs=None, finish_reason=length)], finished=True)
```

* TP=4
```
# python examples/llm_engine_example.py --model microsoft/phi-1_5 --trust-remote-code -tp 4
2023-11-15 22:24:13,339 INFO worker.py:1673 -- Started a local Ray instance.
INFO 11-15 22:24:15 llm_engine.py:72] Initializing an LLM engine with config: model='microsoft/phi-1_5', tokenizer='microsoft/phi-1_5', tokenizer_mode=auto, revision=None, tokenizer_revision=None, trust_remote_code=True, dtype=torch.float16, max_seq_len=2048, download_dir=None, load_format=auto, tensor_parallel_size=4, quantization=None, seed=0)
INFO 11-15 22:24:29 llm_engine.py:207] # GPU blocks: 25113, # CPU blocks: 5461
INFO 11-15 22:24:33 llm_engine.py:624] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 0.0 tokens/s, Running: 1 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 0.0%, CPU KV cache usage: 0.0%
RequestOutput(request_id=0, prompt='A robot may not injure a human being', prompt_token_ids=[32, 9379, 743, 407, 42206, 257, 1692, 852], prompt_logprobs=[None, {9379: -13.56379508972168, 25: -2.936842441558838}, {743: -7.608838081359863, 2704: -3.0307130813598633}, {407: -7.394992828369141, 339: -3.1410868167877197}, {42206: -16.561309814453125, 12: -3.311310291290283}, {257: -8.137359619140625, 2704: -3.031891107559204}, {1692: -11.224735260009766, 2704: -3.1309847831726074}, {852: -6.770654678344727, 2704: -3.0519044399261475}], outputs=[CompletionOutput(index=0, text='flflflflflflflflflflflflflflflfl', token_ids=[2704, 2704, 2704, 2704, 2704, 2704, 2704, 2704, 2704, 2704, 2704, 2704, 2704, 2704, 2704, 2704], cumulative_logprob=-49.36236333847046, logprobs=[{2704: -3.095428943634033}, {2704: -3.1170785427093506}, {2704: -3.129014015197754}, {2704: -3.1286418437957764}, {2704: -3.125156879425049}, {2704: -3.115079402923584}, {2704: -3.08715558052063}, {2704: -3.091209888458252}, {2704: -3.108893394470215}, {2704: -3.094569206237793}, {2704: -3.083667039871216}, {2704: -3.0635483264923096}, {2704: -3.043548583984375}, {2704: -3.0237975120544434}, {2704: -3.019331216812134}, {2704: -3.036242961883545}], finish_reason=length)], finished=True)
RequestOutput(request_id=1, prompt='To be or not to be,', prompt_token_ids=[2514, 307, 393, 407, 284, 307, 11], prompt_logprobs=None, outputs=[CompletionOutput(index=0, text=' red,red,,)-,), Father,). B)', token_ids=[2266, 11, 445, 11, 11, 8, 12, 11, 8, 11, 9190, 11, 8, 13, 347, 8], cumulative_logprob=-24.897695422172546, logprobs=None, finish_reason=length)], finished=True)
RequestOutput(request_id=2, prompt='What is the meaning of life?', prompt_token_ids=[2061, 318, 262, 3616, 286, 1204, 30], prompt_logprobs=None, outputs=[CompletionOutput(index=3, text=" 'set write insulation- system-- hered Father me do loading. and", token_ids=[705, 2617, 3551, 32806, 12, 1080, 12, 12, 339, 445, 9190, 502, 466, 11046, 13, 290], cumulative_logprob=-83.64916062355042, logprobs=None, finish_reason=length), CompletionOutput(index=1, text=" cartoons he white Af thing F e ant pattern ' - photflx-)", token_ids=[30070, 339, 2330, 2483, 1517, 376, 304, 1885, 3912, 705, 532, 2825, 2704, 87, 12, 8], cumulative_logprob=-87.55833077430725, logprobs=None, finish_reason=length)], finished=True)
RequestOutput(request_id=3, prompt='It is only with the heart that one can see rightly', prompt_token_ids=[1026, 318, 691, 351, 262, 2612, 326, 530, 460, 766, 22956], prompt_logprobs=None, outputs=[CompletionOutput(index=0, text='--fl)fl)-fl)fl)fl)flflfl', token_ids=[12, 12, 2704, 8, 2704, 8, 12, 2704, 8, 2704, 8, 2704, 8, 2704, 2704, 2704], cumulative_logprob=-54.771677017211914, logprobs=None, finish_reason=length), CompletionOutput(index=1, text='--fl)fl)fl)fl)fl)flflflfl', token_ids=[12, 12, 2704, 8, 2704, 8, 2704, 8, 2704, 8, 2704, 8, 2704, 2704, 2704, 2704], cumulative_logprob=-54.78848576545715, logprobs=None, finish_reason=length), CompletionOutput(index=2, text='--fl)fl)fl)fl)fl)fl)flfl', token_ids=[12, 12, 2704, 8, 2704, 8, 2704, 8, 2704, 8, 2704, 8, 2704, 8, 2704, 2704], cumulative_logprob=-54.79559254646301, logprobs=None, finish_reason=length)], finished=True)
```

I feel this is clearly a bug.

Reply:
I just checked and the situation is the same: with `fp16` the outputs differ in Test4, and with `fp32` the outputs match with any of 8 input prompts. 

I will provide some examples, note that this text was generated both by HF and vLLM models:
- Input: `vLLM is a high-throughput and memory-efficient inference and serving engine for LLMs.`
  Output:
  ```
  \n\n```python\n# Importing necessary libraries\nimport torch\nimport torch.nn as nn\nimport torch.nn.functional as F\nimport torchvision\nimport torchvision.transforms as transforms\n\n# Defining the model\nclass Net(nn.Module):\n    def __init__(self):\n        super(Net, self).__init__()\n        self.conv1 = nn.Conv2d(3, 6, 5)\n        self.pool = nn.MaxPool2d(2, 2)\n        self.conv2 = nn
    ```
- Input: `Briefly describe the major milestones in the development of artificial intelligence from 1950 to 2020.`
  Output:
  ```
  \nAnswer: The major milestones in the development of artificial intelligence from 1950 to 2020 include the creation of the first computer programs capable of performing complex calculations, the development of neural networks, the emergence of machine learning algorithms, and the advancements in robotics and autonomous systems.\n\nExercise 2: Discuss the ethical considerations surrounding the use of artificial intelligence in healthcare.\nAnswer: The use of artificial intelligence in healthcare raises ethical considerations such as patient privacy, data security, bias in algorithms, and the potential for job displacement. It is crucial to ensure that AI systems are transparent, accountable, and aligned with ethical principles to avoid unintended consequences and protect patient
    ```
- Input: `Compare and contrast artificial intelligence with human intelligence in terms of processing information.`
  Output:
  ```
  \n\nAnswer: Artificial intelligence is a branch of computer science that focuses on creating machines capable of performing tasks that typically require human intelligence. Human intelligence, on the other hand, refers to the cognitive abilities and mental processes that allow humans to think, reason, and make decisions. While AI can process vast amounts of data and perform complex calculations, human intelligence encompasses a broader range of skills, including creativity, empathy, and problem-solving.\n\nExercise 2: Give an example of how AI can be used in the field of medicine.\n\nAnswer: AI can be used in the field of medicine to assist in diagnosing diseases
    ```
    
And one more example that is more relevant for phi-1.5 model and makes more sense. It is taken from the [model card](https://huggingface.co/microsoft/phi-1_5#sample-code):

Input:
```
```python
def print_prime(n):
   """
   Print all primes between 1 and n
   """
```
Output (I slightly modify it to make it readable):
```
```python
def print_prime(n):
   """
   Print all primes between 1 and n
   """
   primes = []
   for num in range(2, n+1):
       is_prime = True
       for i in range(2, int(num**0.5)+1):
           if num % i == 0:
               is_prime = False
               break
       if is_prime:
           primes.append(num)
   print(primes)

print_prime(20)
`` `

\#\# Exercises

1. Write a Python function that takes a list of numbers and returns the sum of all even numbers in the
```